junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertLinesMatch.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.jupiter.api;
+
+import static java.lang.String.format;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.platform.commons.util.Preconditions.condition;
+import static org.junit.platform.commons.util.Preconditions.notNull;
+
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * {@code AssertLinesMatch} is a collection of utility methods that support asserting
+ * lines of {@link String} equality or {@link java.util.regex.Pattern}-match in tests.
+ *
+ * @since 5.0
+ */
+class AssertLinesMatch {
+
+	static void assertLinesMatch(List<String> expectedLines, List<String> actualLines) {
+		notNull(expectedLines, "expectedLines must not be null");
+		notNull(actualLines, "actualLines must not be null");
+
+		// trivial case: same list instance
+		if (expectedLines == actualLines) {
+			return;
+		}
+
+		int expectedSize = expectedLines.size();
+		int actualSize = actualLines.size();
+
+		// trivial case: when expecting more than actual lines available, something is wrong
+		if (expectedSize > actualSize) {
+			// use standard assertEquals(Object, Object, message) to let IDEs present the textual difference
+			String expected = String.join(System.lineSeparator(), expectedLines);
+			String actual = String.join(System.lineSeparator(), actualLines);
+			assertEquals(expected, actual, "expected " + expectedSize + " lines, but only got " + actualSize);
+			fail("should not happen as expected != actual was asserted");
+		}
+
+		// simple case: both list are equally sized, compare them line-by-line
+		if (expectedSize == actualSize) {
+			boolean allOk = true;
+			for (int i = 0; i < expectedSize; i++) {
+				if (matches(expectedLines.get(i), actualLines.get(i))) {
+					continue;
+				}
+				allOk = false;
+				break;
+			}
+			if (allOk) {
+				return;
+			}
+		}
+
+		assertLinesMatchWithFastForward(expectedLines, actualLines);
+	}
+
+	private static void assertLinesMatchWithFastForward(List<String> expectedLines, List<String> actualLines) {
+		Deque<String> expectedDeque = new LinkedList<>(expectedLines);
+		Deque<String> actualDeque = new LinkedList<>(actualLines);
```

Comment:
Is there any particular reason that `LinkedList`s were used instead of `ArrayDeque`s (which I believe usually have better performance)?